cli/upgrade.rs

```diff
@@ -0,0 +1,189 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+
+//! This module provides feature to upgrade deno executable
+//!
+//! At the moment it is only consumed using CLI but in
+//! the future it can be easily extended to provide
+//! the same functions as ops available in JS runtime.
+
+extern crate flate2;
+use crate::fs::write_file;
+use crate::futures::FutureExt;
+use crate::{
+  http_util::{fetch_once, FetchOnceResult},
+  version, ErrBox,
+};
+use flate2::write::GzDecoder;
+use regex::Regex;
+use reqwest::{redirect::Policy, Client};
+use std::env::current_exe;
+use std::fs::{remove_file, rename};
+use std::future::Future;
+use std::io::prelude::*;
+use std::path::Path;
+use std::pin::Pin;
+use std::string::String;
+use url::Url;
+
+lazy_static! {
+  static ref LATEST_VERSION_URL: String =
+    "https://github.com/denoland/deno/releases/latest".to_string();
+  static ref EXEC_DOWNLOAD_URL: String =
+    "https://github.com/denoland/deno/releases/download/v".to_string();
+  static ref REGEX_STRING: String = r#"v([^\?]+)?""#.to_string();
+  static ref DENO_EXEC_TEMP_NAME: String = "deno_temp".to_string();
+}
+
+#[cfg(windows)]
+const EXEC_FILE_NAME: &str = "deno_win_x64.zip";
+#[cfg(macos)]
+const EXEC_FILE_NAME: &str = "deno_osx_x64.gz";
+#[cfg(unix)]
+const EXEC_FILE_NAME: &str = "deno_linux_x64.gz";
+
+struct ErrorMsg(String);
+
+impl ErrorMsg {
+  fn to_err_box(&self) -> ErrBox {
+    ErrBox::from(std::io::Error::new(
+      std::io::ErrorKind::Other,
+      self.0.clone(),
+    ))
+  }
+}
+
+/// Asynchronously updates deno executable to greatest version
+/// if greatest version is available.
+pub async fn exec_upgrade() -> Result<(), ErrBox> {
+  let client = Client::builder().redirect(Policy::none()).build()?;
+  println!("Checking for latest version");
+  let body = client
+    .get(Url::parse(&LATEST_VERSION_URL)?)
+    .send()
+    .await?
+    .text()
+    .await?;
+  let checked_version = find_version(&body)?;
+  if is_latest_version_greater(&version::DENO.to_string(), &checked_version) {
```

Comment:
I hope the `upgrade` command can also take an e.g. `--dry-run/--check` flag that only checks if the current install version is latest.